### PR TITLE
Improve data consistency for IMap and ICache iterators

### PR DESCRIPTION
**Pull request by mmedenjak**
Improves consistency of returned data for IMap and ICache iterators
in the face of concurrent mutation of map and cache data and migration
of data between members.

The new contract for the iterator:
- keys which have been deleted or added since the iteration started 
must be observed at most once, other keys must be observed exactly once
- the next method may return an entry with a null value. This can happen
when prefetchValues is false and the key was present when the entry 
batch was fetched during the consumption of this batch, the entry was 
removed.

The idea is described here:
https://hazelcast.atlassian.net/wiki/spaces/PM/pages/498663502/Support+IMap+iterations+with+concurrent+mutation under `Having the iterator keep a collection of table_index, table_size tuples`

Fixes: hazelcast/hazelcast#11106
EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3151

### CODE REVIEW

**Comment by devozerov** [comment]
I went through the whole flow. The code looks good in general. But my main concern is the potential performance degradation of iteration performance. The principal observation is that table resize is a rare event. We should try to optimize for this case. E.g. a single comparison of expected and actual sizes should be enough to bypass all other newly added operations, and add KV pairs to the batch safely. Likewise, a single `IterationPointer` should be enough for the 99% cases, instead of an array. 

Other than that, this will help a lot e.g. SQL engine to provide more predictable results.

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheRecordStore.java [hunk 1]
```diff
@@ -434,22 +435,38 @@
...
+     * by receiving the {@code tableIndex} of the last item. The index can
+     * represent a bucket with multiple items and in this case the returned
+     * object will contain all items in that bucket, regardless if we exceed
+     * the requested {@code size}.
      *
-     * @param tableIndex initial table index.
-     * @param size       maximum key set size.
-     * @return {@link CacheKeyIterationResult} which wraps keys and last tableIndex.
+     * @param pointers the pointers defining the state of iteration
+     * @param size     the minimal count of returned items
```
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheRecordStore.java]: Yes, you're right.

**Comment by viliam-durina** [comment]
A big thanks from the Jet team and users!

**Pull request merged**